// @salesforce/templates/lib/service/templateService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateService = void 0;
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const path = require("path");
const yeoman = require("yeoman-environment");
const i18n_1 = require("../i18n");
const utils_1 = require("../utils");
const types_1 = require("../utils/types");
/**
 * Template Service
 */
class TemplateService {
    constructor(cwd = process.cwd()) {
        this.adapter = new utils_1.ForceGeneratorAdapter();
        // @ts-ignore the adaptor doesn't fully implement yeoman's adaptor yet
        this.env = yeoman.createEnv(undefined, { cwd }, this.adapter);
    }
    /**
     * Get an instance of TemplateService
     * @param cwd cwd of current yeoman environment. CLI: don't need to set explicitly. VS Code: it's typically the root workspace path
     */
    static getInstance(cwd) {
        if (!TemplateService.instance) {
            TemplateService.instance = new TemplateService(cwd);
        }
        else if (cwd) {
            TemplateService.instance.cwd = cwd;
        }
        return TemplateService.instance;
    }
    /**
     * Getting cwd of current yeoman environment
     */
    get cwd() {
        return this.env.cwd;
    }
    /**
     * Setting cwd of current yeoman environment
     * In VS Code, it's typically the root workspace path
     */
    set cwd(cwd) {
        this.env.cwd = cwd;
    }
    /**
     * Look up package version of @salesforce/templates package to supply a default API version
     */
    static getDefaultApiVersion() {
        const packageJsonPath = path.join('..', '..', 'package.json');
        const versionTrimmed = require(packageJsonPath).version.trim();
        return `${versionTrimmed.split('.')[0]}.0`;
    }
    /**
     * Create using templates
     * @param templateType template type
     * @param templateOptions template options
     */
    create(templateType, templateOptions) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const generatorClass = types_1.TemplateType[templateType]
                .toString()
                .charAt(0)
                .toLowerCase() +
                types_1.TemplateType[templateType].toString().slice(1) +
                'Generator';
            const generatorNamespace = `@salesforce/${generatorClass}`;
            let generator = this.env.get(generatorNamespace);
            if (!generator) {
                generator = (yield Promise.resolve().then(() => require(`../generators/${generatorClass}`))).default;
                const generatorPackagePath = path.join(__dirname, '..', '..');
                this.env.registerStub(generator, generatorNamespace, generatorPackagePath);
            }
            this.adapter.log.clear();
            return new Promise((resolve, reject) => {
                this.env.run(generatorNamespace, templateOptions, err => {
                    if (err) {
                        reject(err);
                    }
                    const outputDir = path.resolve(this.cwd, templateOptions.outputdir);
                    const created = this.adapter.log.getCleanOutput();
                    const rawOutput = i18n_1.nls.localize('RawOutput', [
                        outputDir,
                        this.adapter.log.getOutput()
                    ]);
                    const result = {
                        outputDir,
                        created,
                        rawOutput
                    };
                    resolve(result);
                });
            });
        });
    }
}
exports.TemplateService = TemplateService;
