// @salesforce/plugin-source/lib/commands/force/source/convert.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Convert = void 0;
const os = require("os");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const source_deploy_retrieve_1 = require("@salesforce/source-deploy-retrieve");
const ts_types_1 = require("@salesforce/ts-types");
const sourceCommand_1 = require("../../../sourceCommand");
const convertResultFormatter_1 = require("../../../formatters/convertResultFormatter");
const componentSetBuilder_1 = require("../../../componentSetBuilder");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-source', 'convert');
class Convert extends sourceCommand_1.SourceCommand {
    async run() {
        await this.convert();
        this.resolveSuccess();
        return this.formatResult();
    }
    async convert() {
        const paths = [];
        const { sourcepath, metadata, manifest, rootdir } = this.flags;
        if (sourcepath) {
            paths.push(...sourcepath);
        }
        // rootdir behaves exclusively to sourcepath, metadata, and manifest... to maintain backwards compatibility
        // we will check here, instead of adding the exclusive option to the flag definition so we don't break scripts
        if (rootdir && !sourcepath && !metadata && !manifest) {
            // only rootdir option passed
            paths.push(rootdir);
        }
        // no options passed, convert the default package (usually force-app)
        if (!sourcepath && !metadata && !manifest && !rootdir) {
            paths.push(this.project.getDefaultPackage().path);
        }
        this.componentSet = await componentSetBuilder_1.ComponentSetBuilder.build({
            sourceapiversion: await this.getSourceApiVersion(),
            sourcepath: paths,
            manifest: manifest && {
                manifestPath: this.getFlag('manifest'),
                directoryPaths: this.getPackageDirs(),
            },
            metadata: metadata && {
                metadataEntries: this.getFlag('metadata'),
                directoryPaths: this.getPackageDirs(),
            },
        });
        const converter = new source_deploy_retrieve_1.MetadataConverter();
        this.convertResult = await converter.convert(this.componentSet, 'metadata', {
            type: 'directory',
            outputDirectory: this.getFlag('outputdir'),
            packageName: this.getFlag('packagename'),
            genUniqueDir: false,
        });
    }
    resolveSuccess() {
        if (!ts_types_1.getString(this.convertResult, 'packagePath')) {
            this.setExitCode(1);
        }
    }
    formatResult() {
        const formatter = new convertResultFormatter_1.ConvertResultFormatter(this.logger, this.ux, this.convertResult);
        if (!this.isJsonOutput()) {
            formatter.display();
        }
        return formatter.getJson();
    }
}
exports.Convert = Convert;
Convert.description = messages.getMessage('description');
Convert.examples = messages.getMessage('examples').split(os.EOL);
Convert.requiresProject = true;
Convert.flagsConfig = {
    rootdir: command_1.flags.directory({
        char: 'r',
        description: messages.getMessage('flags.rootdir'),
    }),
    outputdir: command_1.flags.directory({
        default: './',
        char: 'd',
        description: messages.getMessage('flags.outputdir'),
    }),
    packagename: command_1.flags.string({
        char: 'n',
        description: messages.getMessage('flags.packagename'),
    }),
    manifest: command_1.flags.string({
        char: 'x',
        description: messages.getMessage('flags.manifest'),
    }),
    sourcepath: command_1.flags.array({
        char: 'p',
        description: messages.getMessage('flags.sourcepath'),
        exclusive: ['manifest', 'metadata'],
    }),
    metadata: command_1.flags.array({
        char: 'm',
        description: messages.getMessage('flags.metadata'),
        exclusive: ['manifest', 'sourcepath'],
    }),
};
